File: examples/go/cmd/server/dex/money_transfer_controller.go
// Copyright (c) 2022-2026 Super Durable, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

package dex

import (
	"net/http"
	"strconv"

	"github.com/gin-gonic/gin"
	"github.com/superdurable/dex/examples/go/workflows"
	"github.com/superdurable/dex/examples/go/workflows/moneytransfer"
	sdk "github.com/superdurable/dex/sdk-go/dex"
)

type moneyTransferController struct {
	client *sdk.Client
}

func newMoneyTransferController(client *sdk.Client) *moneyTransferController {
	return &moneyTransferController{client: client}
}

func (controller *moneyTransferController) registerRoutes(router *gin.Engine) {
	router.GET("/moneytransfer/start", controller.start)
}

func (controller *moneyTransferController) start(request *gin.Context) {
	amount, err := strconv.Atoi(request.Query("amount"))
	if err != nil {
		request.JSON(http.StatusBadRequest, gin.H{"error": "amount must be an integer"})
		return
	}
	flowID := newFlowID("money-transfer")
	input := moneytransfer.TransferRequest{
		FromAccount: request.Query("fromAccount"),
		ToAccount:   request.Query("toAccount"),
		Amount:      amount,
		Notes:       request.Query("notes"),
	}
	startFlow(request, controller.client, workflows.MoneyTransfer, flowID, input)
}

File: examples/java/src/main/java/io/superdurable/dex/products/moneytransfer/MoneyTransferController.java
/*
 * Copyright (c) 2022-2026 Super Durable, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.superdurable.dex.products.moneytransfer;

import io.superdurable.dex.Client;
import io.superdurable.dex.shared.ExampleFlows;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.util.LinkedHashMap;
import java.util.Map;

@RestController
@RequestMapping("/products/money-transfer")
public class MoneyTransferController {
    private final Client client;
    private final MoneyTransferFlow flow;

    public MoneyTransferController(
            final Client client,
            final MoneyTransferFlow flow) {
        this.client = client;
        this.flow = flow;
    }

    @GetMapping("/start")
    public ResponseEntity<Map<String, String>> start(
            @RequestParam final String fromAccount,
            @RequestParam final String toAccount,
            @RequestParam final int amount,
            @RequestParam(defaultValue = "") final String notes) {
        final String flowId = "money-transfer-" + System.nanoTime();
        final TransferRequest request = new TransferRequest(
                fromAccount,
                toAccount,
                amount,
                notes);
        final String runId = client.startFlow(flow, flowId, request, ExampleFlows.startOptions());
        final Map<String, String> response = new LinkedHashMap<String, String>();
        response.put("flowID", flowId);
        response.put("runID", runId);
        return ResponseEntity.ok(response);
    }
}
